### PR TITLE
Add GNOME 40 support in the extensions prefs window

### PR DESCRIPTION
**Pull request by frandieguez**
Hi, 

please consider merging this PR to add support to GNOME 40 in the preferences page of this extension.

There are some parts that are visually more appealing but there are some rough parts that I'm not particularly happy about.

image

Regards,
Fran

### CODE REVIEW

**Comment by frandieguez** [comment]
In addition I've added 
- Translations for Galician language
- Small Makefile to simpliify build and installation steps

**Comment by maoschanz** [comment]
that's not gnome 3.40 support, that's an entire redesign. I'm ok with gnome 3.40 support, not with a redesign:

- scrolling even a single millimetre in this window is unacceptable, not only because it's very annoying, but also because of the poor discoverability it creates, and because scrolling on combobox or spinbuttons widgets change their values
- libhandy-looking rows with frames are trendy but they put a considerable distance between an item and its label, they're [not what the HIG recommend](https://developer.gnome.org/hig/stable/visual-layout.html.en#label-alignment)

this is a matter of usability, not appeal.

If they removed the possibility to have a stack switcher in the headerbar, then the corresponding code should be deleted, not commented, and the "about" should be somehow moved to a classic GtkAboutDialog 

(also please don't reindent files, especially if it's to replace normal tabs with bunch of spaces)

Concerning the makefile, i appreciate the idea but i'm not sure it's correct: where is $INSTALL_DIR defined?

**Comment by frandieguez** [comment]
First of all, thanks for your quick comments and for your extension.

- I can restore the layout into GtkStack pages in order to remove the scrolling. 
- About the libhandy-looking, I find it an improvement in usability against, the previous layout. About being "and entire redesign" I'm only updating the extension, which I use in a daily basis, to being able to use GNOME 40 . Not sure if you will update in the near future, but would like to use it.

If you find that the changes proposed will not be landed, no problem my own needs are fulfilled so I will wait for your official release. 
Notice that I'm not a GTK developer, and this is the first time that I touch anything related with it. For me is a huge effort to make a simple change cause I have to see other examples, read documentation, etc., and I'm not particularly fond on propose changes that could not be landed.

**Comment by frandieguez** [comment]
I forgot answering about the makefile. This is a quick version so I will update it to support the INSTALL_DIR

**Comment by frandieguez** [comment]
Another thing. Not sure if HIG is obsolete but every major GNOME application is migrating this UX to this kind of frames libhandy-like

https://gitlab.gnome.org/Teams/Design/os-mockups/-/blob/master/app-preferences/app-preferences.png

**Comment by frandieguez** [comment]
Makefile improved with support for standard PREFIX variable, so now in order to install system-wide:
`PREFIX=/usr/share/gnome-shell/extensions/ make install`

**Comment by frandieguez** [comment]
Ok, I've got the prefs panel converted back to GtkStack, but I have some problems that maybe you can help me with. 

I replace de GtkStackSwitcher inside the headerbar but that makes window buttons disappear. I'm just taking ideas from other extensions but not sure if I'm doing well.

Peek 12-03-2021 17-38

**Comment by frandieguez** [comment]
I have changed the GtkStack with a GtkNotebook with gives the prefs panel tabs, so no more scrolls

No plans to do more changes, as I'm not sure if it they will land.

Hope you find these changes useful. And please, don't just me too much for them, as I'm a newbie in Gtk

**Comment by AlexWayfer** [comment]
(you can now add "resolves #151" to the PR description to auto-close a new issue)

**Review by EgidioCaprino — APPROVED** [review]

**Comment by azisnaufal** [comment]
is there any update? will this PR wil get merged?

**Comment by wbolster** [comment]
i would :heart_eyes: to see this merged and released as well. @maoschanz please help the poor souls on gnome 40 out :pray: :upside_down_face:

in the mean time, people can restore basic functionality by editing the metadata file at`$HOME/.local/share/gnome-shell/extensions/emoji-selector@maestroschan.fr/metadata.json`
and adding `"40"` to the `shell-version` list, so that it looks like this:

```
...
    "3.36",
    "3.38",
    "40"
  ],
...
```

**Comment by edevars** [comment]
Please merge this pull request. I downloaded it and works fine. :)

**Comment by avanishsubbiah** [comment]
I can't modify the keyboard shortcut with the new settings menu. Clicking on it doesn't allow for the text to be edited at all.

**Comment by EgidioCaprino** [comment]
That should be fixed once this PR gets merged. In the meantime you can do [this](https://github.com/maoschanz/emoji-selector-for-gnome/pull/149#issuecomment-823044494).

**Comment by avanishsubbiah** [comment]
Ah, besides that it's great. I was able to work around it by typing out the new keyboard shortcut elsewhere and copying and pasting it into the text box.

**Comment by kajlando** [comment]
Someone please fork this extension. The autor doesn't seem to care.

**Comment by kajlando** [comment]
Delete emoji-selector@maestroschan.fr from ~/.local/share/gnome-shell/extensions
Download source from frandieguez/emoji-selector-for-gnome, extract, then install with ./install.sh

**Comment by kajlando** [comment]
Delete emoji-selector@maestroschan.fr from ~/.local/share/gnome-shell/extensions
Download source from frandieguez/emoji-selector-for-gnome, extract, then install with ./install.sh

**Comment by AlexWayfer** [comment]
@maoschanz sorry, but can we get your attention here? ❤️

**Comment by Queatz** [comment]
gahhh I miss having this extension so much

**Comment by AlexWayfer** [comment]
@frandieguez there are conflicts by the way.

**Comment by daPhipz** [comment]
In #156, he wrote:

> the fork compatible with GNOME 40 already exists, there is a pull request opened with it. But it's not compatible with any older GNOME Shell version including the one i use, so i can't merge it yet

**Comment by frandieguez** [comment]
I'll tr to update my PR. Hope it finally lands

**Review by EgidioCaprino — APPROVED** [review]

**Comment by EgidioCaprino** [comment]
@frandieguez could you please resolve the conflicts?

**Comment by jacksongoode** [comment]
@frandieguez We need you 🥺

**Comment by the-loudspeaker** [comment]
I see no conflicts on the branch. yayy!! Hopefully it lands now @maoschanz ?

File: emoji-selector@maestroschan.fr/metadata.json
Comment (shilangyu) [inline emoji-selector@maestroschan.fr/metadata.json:9]: Sorry for a nitpick, but it seems like it needs a reformat

**Comment by the-loudspeaker** [comment]
^ @frandieguez Just a gentle reminder. :)

File: emoji-selector@maestroschan.fr/metadata.json
Comment (frandieguez) [reply emoji-selector@maestroschan.fr/metadata.json]: This is fixed in my master branch and included in this PR.

**Comment by frandieguez** [comment]
I don't know if after a gnome cycle this pr will get merged.  I'm willing, hope, and pray that this will finally gets merged but have no news from the mantainer.

I use this extension on a daily basis, and for now it is working perfectly.

**Comment by the-loudspeaker** [comment]
yay! finally without any conflicts. Hopefully it gets merged now. :)
cc: @shilangyu

**Comment by stephane-klein** [comment]
I don't know if it is a bug related to this pull request, but I can't type key in `keybinding_entry` `GtkEntry` widget:

image

I can't update `<super>e` value.

**Comment by stephane-klein** [comment]
I don't know if it is a bug related to this pull request, any top icon change icon list display.

image

And when I click on the imoji, nothing happens, no copy, picker isn't closed…

File: emoji-selector@maestroschan.fr/prefs.js
Comment (stephane-klein) [inline emoji-selector@maestroschan.fr/prefs.js:182]: @frandieguez **issue:** I have this error:

```
(org.gnome.Shell.Extensions:12542): Gjs-CRITICAL **: 22:59:56.270: JS ERROR: TypeError: this._registerSignals is not a function
buildPrefsWidget/<@/home/stephane/.local/share/gnome-shell/extensions/emoji-selector@maestroschan.fr/prefs.js:184:8
OpenExtensionPrefsAsync/<@resource:///org/gnome/Shell/Extensions/js/extensionsService.js:139:20
asyncCallback@resource:///org/gnome/gjs/modules/core/overrides/Gio.js:115:22
run@resource:///org/gnome/Shell/Extensions/js/dbusService.js:177:20
main@resource:///org/gnome/Shell/Extensions/js/main.js:19:13
run@resource:///org/gnome/gjs/modules/script/package.js:206:19
start@resource:///org/gnome/gjs/modules/script/package.js:190:8
@/usr/share/gnome-shell/org.gnome.Shell.Extensions:1:17
```

Comment (jacksongoode) [reply emoji-selector@maestroschan.fr/prefs.js:182]: @frandieguez is this a breaking error?

**Comment by chmouel** [comment]
Maybe it's me but I tried this PR and I could not get this to work without having to resort to this (i guess ugly) patch, my knowledge of extensions and javascript is pretty non existent so that may be completely wrong, but testing your PR with that patch on gnome4.1/arch and it works well enough for me to search and copy emojis correctly ☺️

```diff
diff --git a/emoji-selector@maestroschan.fr/emojiButton.js b/emoji-selector@maestroschan.fr/emojiButton.js
index fdc615b..d26928b 100644
--- a/emoji-selector@maestroschan.fr/emojiButton.js
+++ b/emoji-selector@maestroschan.fr/emojiButton.js
@@ -185,7 +185,7 @@ var EmojiButton = class EmojiButton {
                if (genrable) {
                        temp += GENDERS[Extension.SETTINGS.get_int('gender')];
                }
-               Extension.GLOBAL_BUTTON.searchItem.shiftFor(temp);
+        //Extension.GLOBAL_BUTTON.searchItem.shiftFor(temp);
                return temp;
        }
 };
diff --git a/emoji-selector@maestroschan.fr/emojiCategory.js b/emoji-selector@maestroschan.fr/emojiCategory.js
index c4ee305..908b3c2 100644
--- a/emoji-selector@maestroschan.fr/emojiCategory.js
+++ b/emoji-selector@maestroschan.fr/emojiCategory.js
@@ -176,7 +176,8 @@ var EmojiCategory = class EmojiCategory {
         */
        build() {
                if (this._built) { return; }
-               let ln, container;
+               let ln;
+        let container = new St.BoxLayout();
                for (let i = 0; i < this.emojiButtons.length; i++) {
                        // lines of emojis
                        if (i % this._nbColumns === 0) {
```

**Review by Zorono — COMMENTED** [review]
iam so sorry to say that but this is a very **unstable** edit (most of the features aren't working)... 

----------
* Gnome Version: 40.4.0
* Windowing System: Wayland
* GTK version: 4.4.0
* Adwaita version: 1.0.0
* GStreamer version: 1.19.1
* GJS version: 1.68.4

**Review by Zorono — COMMENTED** [review]
iam so sorry to say that but this is a very **unstable** edit (most of the features aren't working)... 

----------
* Gnome Version: 40.4.0
* Windowing System: Wayland
* GTK version: 4.4.0
* Adwaita version: 1.0.0
* GStreamer version: 1.19.1
* GJS version: 1.68.4

**Comment by maoschanz** [comment]
sorry for not giving any updates, i wasn't motivated at the idea of merging something that would break the extension for basically everybody using an LTS, including me so i couldn't test it at all

my vacations just started so i took a little time to install gnome 41 on another machine to test things, and your version is cool it works well, thank you very much

concerning compatibility there is a solution i just tried (with another of my extension) and i worked:

- move your class to a file named `prefs-gtk4.js` for example
- put the former class in `prefs-gtk3.js`
- in `prefs.js` itself, the function at the bottom should test what's the value of the gnome shell version (put a bogus condition if you're not sure how to do it, i can fix it later) and depending of the version it imports either M̀e.imports.prefs-gtk3.js or M̀e.imports.prefs-gtk4.js and then initialize the object

(no need to test the gtk3 version i'll test it myself after the merge)

File: emoji-selector@maestroschan.fr/prefs.js
Comment (matejmosko) [reply emoji-selector@maestroschan.fr/prefs.js:182]: It's same for me. Clicking on emojis does nothing. Neither copying them in pastebin nor closing the picker.

**Comment by albertvaka** [comment]
Friendly ping :) If this got merged it would make at least 48 people happy, looking at the upvotes :)

**Comment by balazsorban44** [comment]
Workaround:

1. `git clone git@github.com:maoschanz/emoji-selector-for-gnome.git`
2. `cd emoji-selector-for-gnome`
3. `gh pr checkout 149`
4. `./install.sh`
5. logout/login
6. Enjoy! :tada:

**Comment by 3-w-c** [comment]
Bless you @balazsorban44 for the handy howto and the excellent vibes :pray: 

For anyone else who doesn't have the `gh` command line tool:

```
git clone https://github.com/maoschanz/emoji-selector-for-gnome
cd emoji-selector-for-gnome
git remote add frandieguez https://github.com/frandieguez/emoji-selector-for-gnome
git checkout -b pr-149 frandieguez/master
./install.sh
```

(also, here on `gnome-shell-41.2-1`, clicking on groups works OK!)

**Comment by illwieckz** [comment]
Or even (github specific, though):

```
git clone https://github.com/maoschanz/emoji-selector-for-gnome.git
cd emoji-selector-for-gnome
git fetch origin pull/149/head:pr-149
git checkout pr-149
./install.sh
```

Or even (not github specific):

```
git clone https://github.com/maoschanz/emoji-selector-for-gnome.git
cd emoji-selector-for-gnome
git fetch https://github.com/frandieguez/emoji-selector-for-gnome.git master:pr-149
git checkout pr-149
./install.sh
```

Both avoids to add an extra remote.

😃️

But on my end the emojis and tabs are not selectable. 🤷‍♀️️

So my workaround is to always have a gedit window open and use “_right click → insert emoji_” feature and then I copy-paste. 👀️

**Comment by 3-w-c** [comment]
Thanks, even better @illwieckz :clap: I think cloning this repo rather than the fork is probably a better call for most, but yes avoiding the extra remote is great.

I've now seen the "not able to select emojis" problem – it's broken on Xorg, but working on Wayland.

BTW I think there's a keyboard shortcut for the GTK emoji picker you're talking about in Gedit - `Ctrl+Shift+;` (semicolon), also works for me in Evolution, and probably other GTK apps.

**Comment by lyrixx** [comment]
> But on my end the emojis and tabs are not selectable. woman_shrugging️

& 

>I've now seen the "not able to select emojis" problem – it's broken on Xorg, but working on Wayland.

I'm using wayland and I have the same issue (not selectable). I'm using the fork OC :)

**Comment by Syzygianinfern0** [comment]
Does this work?

**Comment by yavko** [comment]
> Does this work on Gnome 42?

Not for me

**Comment by dlecan** [comment]
You need to add `42` to the list of compatible Gnome versions in the file `emoji-selector@maestroschan.fr/metadata.json` to make it work on Gnome 42

**Comment by yavko** [comment]
Oh wow it got merged! 🎉

**Comment by fgrosse** [comment]
I added `42` to the `metadata.json` and managed to install and start the extension successfully but unfortunately it does not show any emojis (also no errors). Does anybody have a quick solution or should I open a new issue for this?

image

Thanks in advance :pray:

**Comment by dispew** [comment]
I really like this extension and I was a bit frustrated that it wasn't working on Gnome 40+. Tried the latest commit [3fe617c](https://github.com/maoschanz/emoji-selector-for-gnome/commit/3fe617c880365b34dcfb23c9c73a9880356a318c) on Pop-OS 21.10 @ Gnome 40.4.0, it was unsuccessful. I forked and made some changes [929b2eb](https://github.com/dispew/emoji-selector-for-gnome/commit/929b2ebb6664f1220b8c6d9a907803c046bcdeed) (also CSS, sorry, was strange on Pop-OS dark theme), works good enough to me. I do believe it need some more changes and code clean-up, so I will not submit a PR.

I hope it helps somehow and sorry for any mistakes. English is not my native language.

image

**Comment by raphink** [comment]
FTR while this extension was broken, I discovered the native way to insert emojis using `Ctrl+;` and `Ctrl+.`

**Comment by lhindir** [comment]
> native way to insert emojis using `Ctrl+;` and `Ctrl+.`

Whoa, thanks. I might be able to drop this extension now. Unfortunately, it appears to be GTK-specific, so it won’t work in Qt or Electron apps, and it doesn’t appear to be enabled in Firefox[^1], either.

[^1]: [1232371 - [UX] Emoji picker for all platforms](https://bugzilla.mozilla.org/show_bug.cgi?id=1232371)

**Comment by yavko** [comment]
> > native way to insert emojis using `Ctrl+;` and `Ctrl+.`
> 
> Whoa, thanks. I might be able to drop this extension now. Unfortunately, it appears to be GTK-specific, so it won’t work in Qt or Electron apps, and it doesn’t appear to be enabled in Firefox[1](#user-content-fn-1-5e2c1fc277d50183a43a3d54d402f7b6), either.
> ## Footnotes
> 
>     1. [1232371 - [UX] Emoji picker for all platforms](https://bugzilla.mozilla.org/show_bug.cgi?id=1232371) [leftwards_arrow_with_hook](#user-content-fnref-1-5e2c1fc277d50183a43a3d54d402f7b6)

Try with ibus

**Comment by DZPM** [comment]
The "native" ways don't work with Firefox & Chrome. We spend most of the time on our browsers, so we need a proper way to add emojis.

As an alternative, I'm using/evaluating:
- **Smile**: https://smile.mijorus.it
- **Emote**: https://github.com/tom-james-watson/Emote

**Comment by raphink** [comment]
The native way works in Chrome and Firefox on my Ubuntu 22.04.

Iirc I had to adjust some environment variables. Not sure if anything else was required:

```bash
GTK_IM_MODULE=ibus
QT_IM_MODULE=ibus
XMODIFIERS=@im=ibus
```

**Comment by AlexWayfer** [comment]
> The "native" ways don't work with Firefox & Chrome. We spend most of the time on our browsers, so we need a proper way to add emojis.
> 
> As an alternative, I'm using/evaluating:
> 
> * **Smile**: https://smile.mijorus.it
> * **Emote**: https://github.com/tom-james-watson/Emote

For me, it works in Chrome, even in Atom Editor. I just was unable to realize what underscored `e` means, I guessed it's some weird tech input enabling by occasional `Ctrl+.`. @raphink helped me to realize that this is just a (weird?) input for emoji auto-completion.

BTW I had problems with Emote installation, and have not seen Smile, but I'll look at them, thank you.

**Comment by raphink** [comment]
Yes, that underscored `e` syntax is quiet weird. Btw, you can also tune its behavior by launching `ibus-setup`, for example so that it matches emojis containing the text you typed instead of the exact match:

**Comment by yavko** [comment]
> Yes, that underscored `e` syntax is quiet weird. Btw, you can also tune its behavior by launching `ibus-setup`, for example so that it matches emojis containing the text you typed instead of the exact match:
> 
> image

There is also a popup menu for that thingy, I think its if you press space twice idk I forgot

**Comment by DZPM** [comment]
> The native way works in Chrome and Firefox on my Ubuntu 22.04.
> 
> Iirc I had to adjust some environment variables. Not sure if anything else was required:
> 
> ```shell
> GTK_IM_MODULE=ibus
> QT_IM_MODULE=ibus
> XMODIFIERS=@im=ibus
> ```

Unlucky me, I can't make it work. The env variables are set, I've rebooted, but still nothing.

Do you have any doc or tutorial that explains the steps in detail?

**Comment by olibre** [comment]
> Do you have any doc or tutorial that explains the steps in detail?

In your `~/.profile` append the following:

```
# The following env. vars enable ibus-emoji using Ctrl + .
export GTK_IM_MODULE=ibus
export QT_IM_MODULE=ibus
export XMODIFIERS=@im=ibus
```

To install *Intelligent Input Bus* see:
https://github.com/ibus/ibus/wiki/ReadMe

To configure Emoji settings:
1. shortcut Alt + F2 then `ibus-setup` or in a terminal, run `ibus-setup`
2. Go to tab Emoji

**Comment by lhindir** [comment]
> Try with ibus

Awesome, thanks. I thought IBus was only for international input, but this emoji feature indeed works across almost all applications I’ve tried. Only Spotify and Steam don’t seem to work, and I never really need emoji input there anyway. It’s also independent of GNOME, so it should work when trying other desktop environments and window managers. This obsoletes this extension for me.

**Pull request merged**
